syntax = "proto3";

package envoy.config.filter.http.localcongestionlimiter_filter.v3;
option go_package = "v3";

import "validate/validate.proto";
import "envoy/config/filter/http/matcher/v3/matcher.proto";
import "envoy/config/filter/http/action/v3/action.proto";
import "envoy/config/filter/http/circuitbreaker_filter/v3/circuitbreaker_filter.proto";
import "envoy/config/route/v3/route_components.proto";

// Localcongestionlimiter filter config
// [#protodoc-title: Localcongestionlimiter Filter]

message LocalCongestionLimiter {
  repeated LocalCongestionLimit congestion_limits = 1;

  repeated Responsepolicy response_policies = 2;

  // Whether the filter is enabled.
  bool filter_enabled = 3;
}

message LocalCongestionLimit {
  // Name for the limit.
  string name = 1;
 
  // A list of actions that are to be applied for this rate limit configuration.
  // Order matters as the actions are processed sequentially and the descriptor
  // is composed by appending descriptor entries in that sequence. If an action
  // cannot append a descriptor entry, no descriptor is generated for the
  // configuration. See :ref:`composing actions
  // <config_http_filters_rate_limit_composing_actions>` for additional documentation.
  repeated route.v3.RateLimit.Action descriptor_components = 2 [(validate.rules).repeated = {min_items: 1}];

  repeated CongestionLimitDescriptor descriptors = 3 [(validate.rules).repeated = {min_items: 1}]; 

  // A name reference to a response_policy
  string response_policy = 4;
}

// Congestion Limit / Rate limit configuration descriptor.
// Mostly from https://github.com/envoyproxy/ratelimit/blob/main/api/ratelimit/config/ratelimit/v3/rls_conf.proto
message CongestionLimitDescriptor {
  // Key of the descriptor.
  string key = 1;

  // Optional value of the descriptor.
  string value = 2;

  // Congestion policy of the descriptor.
  envoy.config.filter.http.circuitbreaker_filter.v3.CBConfig limit = 3;

  // List of sub rate limit descriptors.
  repeated CongestionLimitDescriptor descriptors = 4;

  // Mark the descriptor as shadow. When the value is true, limiter allows requests to the backend.
  bool shadow_mode = 5;
}


message Responsepolicy {

  message MatchAction {
    // Match
    envoy.config.filter.http.matcher.v3.Match match = 2;
    // Action
    envoy.config.filter.http.action.v3.Action action = 3;
  }

  // Name for the policy.
  string name = 1;

  repeated MatchAction responses = 2;
}
